utils/common/src/main/kotlin/EnvironmentVariableFilter.kt

```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.utils.common
+
+/**
+ * An object providing functionality to filter environments that are passed to newly created processes.
+ *
+ * For many tasks, ORT spawns new processes using the [ProcessCapture] class. When creating a new process, the child
+ * process per defaults inherits all environment variables from the parent. This could impose a security risk, for
+ * instance if logic in build scripts could theoretically access sensitive information stored in environment
+ * variables, such as database or service credentials.
+ *
+ * To reduce this list, this object filters the environment variables passed to child processes based on the
+ * following criteria:
+ * - Patterns on variable names can be defined to determine variables with sensitive information. The object provides
+ *   some default patterns to match variables containing strings like "PASS", "USER", "TOKEN", etc.
+ * - There is an allow list to include variables even if they are matched by a pattern.
+ *
+ * So in order to determine whether a specific variable "E" can be passed to a child process, this filter applies the
+ * following steps:
+ * - If E is contained in the allow list, it is included.
+ * - Otherwise, E is included if and only if it is not matched by one of the exclusion patterns.
+ */
+object EnvironmentVariableFilter {
+    /** A list with patterns for variable names that are excluded by default. */
+    val DEFAULT_EXCLUSION_PATTERNS = listOf(
+        "key",
+        "pass",
+        "pwd",
+        "token",
+        "user"
+    )
+
+    /** A list with known variable names that are accepted despite they are matched by exclusion patterns. */
+    val DEFAULT_ALLOW_LIST = setOf(
+        "CARGO_HTTP_USER_AGENT",
+        "COMPOSER_ALLOW_SUPERUSER",
+        "CONAN_LOGIN_ENCRYPTION_KEY",
+        "CONAN_LOGIN_USERNAME",
+        "CONAN_PASSWORD",
+        "CONAN_USERNAME",
+        "CONAN_USER_HOME",
+        "CONAN_USER_HOME_SHORT",
+        "DOTNET_CLI_CONTEXT_ANSI_PASS_THRU",
+        "GIT_ASKPASS",
+        "GIT_HTTP_USER_AGENT",
+        "GRADLE_USER_HOME",
+        "HACKAGE_USERNAME",
+        "HACKAGE_PASSWORD",
+        "HACKAGE_KEY",
+        "PWD",
+        "USER",
+        "USERPROFILE"
+    )
+
+    /** Stores the current exclusion patterns used by this filter. */
+    private var exclusionPatterns = DEFAULT_EXCLUSION_PATTERNS.toPatterns()
+
+    /** Stores the current allow list used by this filter. */
+    private var allowList = DEFAULT_ALLOW_LIST
+
+    /**
+     * Initialize this filter from the given [exclusionPatterns] and [allowList]. The [exclusionPatterns] are
+     * converted to regular expressions, so that they can be matched anywhere in variable names, ignoring case.
+     */
+    fun init(
```

Comment:
Please don't call this function `init`. That's way too easy to confuse with the `init {}` block. Maybe `reset`?

Maybe this `object` should become a `class` after all, with a factory method in a `companion object` that creates a default instance?

Reply:
Renamed the function to reset. As discussed during a call, before converting the `object` to a `class`, we wait for the introduction of dependency injection facilities.

Reply:
See https://github.com/oss-review-toolkit/ort/pull/6037 for some early experiments.

Reply:
Alternatively, this PR could be merged as is, and the conversion from an `object` to a `class` is done afterwards. I guess, the best way forward depends on the time you need for your experiments. As I said, this PR is not that urgent for us, but if it remains open for a longer time, there is always the risk for conflicts.

Reply:
> Alternatively, this PR could be merged as is, and the conversion from an `object` to a `class` is done afterwards.

Yes, I'd be fine with this, to not put me and my experiments in the position of a bottleneck. I'll review this PR once more from that perspective. But I'd actually appreciate also a review from another Bosch contributor, as the feature seems to be mostly wanted from your side.

Reply:
Thanks, fair enough. And I just noticed that my latest renaming has caused a compilation failure.